homeassistant/core.py

```diff
@@ -1451,6 +1450,16 @@ def async_fire(
 
         This method must be run in the event loop.
         """
+        if (
```

Comment:
Since most of the events (state events) come in via _async_fire, I think in this case we might even consider that it's worth the performance tradeoff here to always check even without debug mode since it's frequently the place where a CC gets this wrong

Reply:
We can rename _async_fire to async_fire_internal and change all known safe high volume usage to use that

Then we can make the length validation a separate function and call it in fire as well and update fire to use the internal function since we know it's passing through a call soon threadsafe

Finally we turn the threading check on in async_fire since it would only be rarely called from integrations now as all our internal usage would avoid the check.

Reply:
At least on my production, `_async_fire` is called orders of magnitude ore than `async_fire` so this seems like the overall performance impact will be minimal and it seems worth that tradeoff 

<img width="591" alt="Screenshot 2024-04-23 at 11 57 48 AM" src="https://github.com/home-assistant/core/assets/663432/2b36b6c5-4f52-4d05-a82b-f6e6758d700c">

Reply:
Started working on that in https://github.com/home-assistant/core/pull/116027